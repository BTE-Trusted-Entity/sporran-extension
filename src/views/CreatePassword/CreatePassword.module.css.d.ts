export const __esModule: true;
export const container: string;
export const heading: string;
export const subline: string;
export const criteriaHeading: string;
export const criteria: string;
export const pass: string;
export const inputLine: string;
export const input: string;
export const errorTooltip: string;
export const cancel: string;
export const submit: string;
export const complexity: string;
export const complexityNone: string;
export const complexityPoor: string;
export const complexityMedium: string;
export const complexityGood: string;
export const complexityOk: string;
export const lock: string;
