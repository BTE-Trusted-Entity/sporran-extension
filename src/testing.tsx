export * from '@testing-library/react';
export { waitForElementToBeRemoved } from '@testing-library/dom';
import { act, render as externalRender, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import dialogPolyfill from 'dialog-polyfill';

import { AccountsProviderMock } from './testing/AccountsProviderMock';

export {
  AccountsProviderMock,
  accountsMock,
} from './testing/AccountsProviderMock';

export { mockBackgroundScript } from './testing/mockBackgroundScript';

export function render(
  ui: Parameters<typeof externalRender>[0],
  options?: Parameters<typeof externalRender>[1],
): ReturnType<typeof externalRender> {
  return externalRender(
    <AccountsProviderMock>
      <MemoryRouter>{ui}</MemoryRouter>
    </AccountsProviderMock>,
    options,
  );
}

export async function waitForTooltipUpdate(): Promise<void> {
  expect(await screen.findByRole('status')).toHaveAttribute(
    'data-popper-escaped',
  );
}

const dialogPromise = Promise.resolve();
(dialogPolyfill.registerDialog as jest.Mock).mockReturnValue(dialogPromise);

export async function waitForDialogUpdate(): Promise<void> {
  await act(() => dialogPromise);
}
