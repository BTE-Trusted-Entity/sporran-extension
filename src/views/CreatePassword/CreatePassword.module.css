.container {
  composes: container from '../../components/View/View.module.css';
}

.heading {
  composes: screenHeadline from '../../components/Typography/Typography.module.css';
}

.subline {
  composes: subline from '../../components/Typography/Typography.module.css';
  margin: 0;
  line-height: 1.25rem;
}

.criteriaHeading {
  composes: importantHeadline from '../../components/Typography/Typography.module.css';
  margin: 0;
  margin-top: 1rem;
}

.criteria {
  composes: errorText from '../../components/Typography/Typography.module.css';
  font-size: 0.75rem;
  color: var(--color-orange);
  list-style: none;
  margin-top: 0.375rem;
  margin-bottom: 0.375rem;
  padding: 0;
  display: flex;
  flex-direction: column;
  align-items: center;
}

.pass {
  color: var(--color-grey-light);
  padding: 0 1.125rem;
  background: url('../../components/images/checkmark-circle.svg') no-repeat
    right 75%/0.75rem;
}

.inputLine {
  display: inline-block;
  position: relative;
}

.input {
  composes: input from '../../components/Input/Input.module.css';
  width: calc(230rem / 16);
  padding-right: 2.5rem;
}

.errorTooltip {
  composes: tooltip from '../../components/ErrorTooltip/ErrorTooltip.module.css';
}

.cancel {
  composes: buttonSecondary from '../../components/Button/Button.module.css';
  margin-right: 1rem;
}

.submit {
  composes: buttonPrimary from '../../components/Button/Button.module.css';
}

.complexity {
  --color-yellow: #fef601;

  width: calc(100px);
  height: calc(100px);
  margin: 1.5rem auto;
  display: grid;
  grid-template-areas: 'self';
}

.complexity::before {
  grid-area: self;
  content: '';
  /* stylelint-disable-next-line  max-line-length */
  clip-path: path(
    'M0,50.2C0,22.5,22.4,0,50,0s50,22.5,50,50.2c0,24.9-18.5,46.3-43,49.8c-2,0.3-3.9-1.1-4.2-3.2 c-0.3-2.1,1.1-4,3.2-4.2c20.8-2.9,36.6-21.1,36.6-42.2C92.5,26.8,73.4,7.7,50,7.7S7.6,26.8,7.6,50.2c0,21.3,15.8,39.5,36.8,42.3 c2,0.3,3.5,2.2,3.2,4.2c-0.3,2.1-2.2,3.5-4.2,3.2C18.6,96.6,0,75.2,0,50.2'
  );
}

.complexityNone::before {
  background: var(--color-orange);
}

.complexityPoor::before {
  background-image: conic-gradient(
    from 180deg at center,
    var(--color-orange),
    var(--color-yellow)
  );
}

.complexityMedium::before {
  background: var(--color-yellow);
}

.complexityGood::before {
  background-image: conic-gradient(
    from 180deg at center,
    var(--color-yellow),
    var(--color-green)
  );
}

.complexityOk::before {
  background: var(--color-green);
}

.lock {
  grid-area: self;
  margin: auto;
  width: calc(57rem / 16);
  height: calc(64rem / 16);
  /* stylelint-disable-next-line  max-line-length */
  clip-path: path(
    'M37.419 24.346v-1.338c0-4.792-3.567-8.805-8.192-9.25-2.564-.112-5.127.612-6.91 2.396a8.782 8.782 0 0 0-2.843 6.408v1.784c-3.4.78-5.963 3.733-5.963 7.356V42.29c0 4.012 3.288 7.133 7.134 7.133h15.492c4.012 0 7.133-3.288 7.133-7.133V31.702c.112-3.623-2.452-6.576-5.851-7.356zm-12.985-5.963c1.17-1.17 2.842-1.616 4.626-1.505 3.009.279 5.35 3.01 5.35 6.242v1.058H22.483v-1.616c0-1.616.724-3.12 1.95-4.18zM40.372 42.4c0 2.397-1.95 4.18-4.18 4.18H20.7c-2.229 0-4.18-1.95-4.18-4.18V31.813c0-2.508 2.063-4.625 4.626-4.625h14.6c2.509 0 4.626 1.95 4.626 4.458v10.755z M28.447 33.15c-.892 0-1.505.614-1.505 1.505v6.019c0 .892.613 1.505 1.505 1.505.891 0 1.504-.613 1.504-1.505v-6.019c0-.891-.613-1.504-1.504-1.504z'
  );
}

.complexityNone .lock,
.complexityPoor .lock {
  background: var(--color-orange);
}

.complexityMedium .lock,
.complexityGood .lock {
  background: var(--color-yellow);
}

.complexityOk .lock {
  background: var(--color-green);
}
