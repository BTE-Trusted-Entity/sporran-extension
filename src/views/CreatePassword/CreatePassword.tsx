import { ChangeEvent, FormEvent, useCallback, useState } from 'react';
import { browser } from 'webextension-polyfill-ts';
import { Link } from 'react-router-dom';

import * as styles from './CreatePassword.module.css';

import { LinkBack } from '../../components/LinkBack/LinkBack';
import { usePasswordType } from '../../components/usePasswordType/usePasswordType';
import { paths } from '../paths';

function hasUpperCase(value: string): boolean {
  return /\p{Upper}/u.test(value);
}

function hasLowerCase(value: string): boolean {
  return /\p{Lower}/u.test(value);
}

function hasBothCases(value: string): boolean {
  return hasUpperCase(value) && hasLowerCase(value);
}

function hasNumber(value: string): boolean {
  return /\p{Number}/u.test(value);
}

function hasOther(value: string): boolean {
  return /[^\p{Upper}\p{Lower}\p{Number}]/u.test(value);
}

const MIN_LENGTH = 12;

function isLong(value: string): boolean {
  return value.length >= MIN_LENGTH;
}

function isNotExample(value: string): boolean {
  return value !== browser.i18n.getMessage('view_CreatePassword_example');
}

function getComplexityClassName(password: string): string {
  const complexity = [
    hasBothCases(password),
    hasNumber(password),
    hasOther(password),
    isLong(password),
    isNotExample(password),
  ].filter(Boolean).length;

  const classNames = [
    styles.complexityNone,
    styles.complexityNone,
    styles.complexityPoor,
    styles.complexityMedium,
    styles.complexityGood,
    styles.complexityOk,
  ];

  return classNames[complexity];
}

interface Props {
  onSuccess: (password: string) => void;
}

export function CreatePassword({ onSuccess }: Props): JSX.Element {
  const t = browser.i18n.getMessage;

  const [password, setPassword] = useState('');
  const modified = password !== '';
  const { passwordType, passwordToggle } = usePasswordType();

  const handleInput = useCallback((event: ChangeEvent<HTMLInputElement>) => {
    setPassword(event.target.value);
  }, []);

  const error =
    modified &&
    [
      !hasBothCases(password) && t('view_CreatePassword_error_cases'),
      !hasNumber(password) && t('view_CreatePassword_error_numbers'),
      !hasOther(password) && t('view_CreatePassword_error_other'),
      !isLong(password) && t('view_CreatePassword_error_length', [MIN_LENGTH]),
      !isNotExample(password) && t('view_CreatePassword_error_example'),
    ].filter(Boolean)[0];

  const handleSubmit = useCallback(
    async (event: FormEvent) => {
      event.preventDefault();

      if (modified && !error) {
        onSuccess(password);
      }
    },
    [error, modified, onSuccess, password],
  );

  function makeClasses(validator: (value: string) => boolean) {
    const pass = validator(password);
    return modified && pass ? styles.pass : undefined;
  }

  return (
    <main className={styles.container}>
      <h1 className={styles.heading}>{t('view_CreatePassword_heading')}</h1>
      <p className={styles.subline}>
        {t('view_CreatePassword_explanation')}{' '}
        {t('view_CreatePassword_example')}
      </p>

      <h2 className={styles.criteriaHeading}>
        {t('view_CreatePassword_criteria')}
      </h2>
      <ul className={styles.criteria}>
        <li className={makeClasses(hasBothCases)}>
          {t('view_CreatePassword_criteria_cases')}
        </li>
        <li className={makeClasses(hasOther)}>
          {t('view_CreatePassword_criteria_other')}
        </li>
        <li className={makeClasses(isLong)}>
          {t('view_CreatePassword_criteria_length')}
        </li>
        <li className={makeClasses(hasNumber)}>
          {t('view_CreatePassword_criteria_numbers')}
        </li>
        <li className={makeClasses(isNotExample)}>
          {t('view_CreatePassword_criteria_example')}
        </li>
      </ul>

      <form onSubmit={handleSubmit}>
        <p className={styles.inputLine}>
          <input
            className={styles.input}
            onInput={handleInput}
            type={passwordType}
            name="password"
            autoComplete="new-password"
            autoFocus
            required
            minLength={MIN_LENGTH}
            aria-label={t('view_CreatePassword_label')}
            placeholder={t('view_CreatePassword_label')}
          />

          {passwordToggle}

          <output className={styles.errorTooltip} hidden={!error}>
            {error}
          </output>
        </p>

        <div
          className={`${styles.complexity} ${getComplexityClassName(password)}`}
        >
          <div className={styles.lock} />
        </div>

        <p>
          <Link to={paths.home} className={styles.cancel}>
            {t('common_action_cancel')}
          </Link>

          <button
            className={styles.submit}
            type="submit"
            disabled={!modified || Boolean(error)}
          >
            {t('view_CreatePassword_CTA')}
          </button>
        </p>
      </form>

      <LinkBack />
    </main>
  );
}
