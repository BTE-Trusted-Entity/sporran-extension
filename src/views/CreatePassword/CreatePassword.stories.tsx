import { Meta } from '@storybook/react';
import { action } from '@storybook/addon-actions';

import { CreatePassword } from './CreatePassword';

export default {
  title: 'Views/CreatePassword',
  component: CreatePassword,
} as Meta;

export function Template(): JSX.Element {
  return <CreatePassword onSuccess={action('onSuccess')} />;
}
