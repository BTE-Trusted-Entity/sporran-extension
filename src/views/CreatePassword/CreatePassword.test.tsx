import userEvent from '@testing-library/user-event';

import { render, screen } from '../../testing/testing';

import { CreatePassword } from './CreatePassword';

const onSuccess = jest.fn();

const props = {
  onSuccess,
};

describe('CreatePassword', () => {
  it('should render', async () => {
    const { container } = render(<CreatePassword {...props} />);
    expect(container).toMatchSnapshot();
  });

  it('should allow the strong password', async () => {
    render(<CreatePassword {...props} />);

    await userEvent.click(screen.getByText('Next Step'));
    expect(onSuccess).not.toHaveBeenCalled();

    await userEvent.type(
      screen.getByLabelText('Please enter your password'),
      'Hello, World!11',
    );
    await userEvent.click(screen.getByText('Next Step'));

    expect(onSuccess).toHaveBeenCalled();
  });

  it('should allow visible and hidden passwords', async () => {
    const { container } = render(<CreatePassword {...props} />);
    expect(container).toMatchSnapshot();

    await userEvent.click(screen.getByLabelText('Show'));
    expect(container).toMatchSnapshot();

    await userEvent.click(screen.getByLabelText('Hide'));
    expect(container).toMatchSnapshot();
  });

  it('should report weak password errors', async () => {
    const { container } = render(<CreatePassword {...props} />);
    const password = screen.getByLabelText('Please enter your password');

    await userEvent.type(password, 'H');
    expect(container).toMatchSnapshot();

    await userEvent.type(password, 'ello');
    expect(container).toMatchSnapshot();

    await userEvent.type(password, '1');
    expect(container).toMatchSnapshot();

    await userEvent.type(password, '!!!!!');
    expect(container).toMatchSnapshot();

    // here the test should not report any more errors
    await userEvent.type(password, 'f');
    expect(container).toMatchSnapshot();
  });
});
