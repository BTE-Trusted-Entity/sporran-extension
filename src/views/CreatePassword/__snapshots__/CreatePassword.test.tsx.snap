// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreatePassword should allow visible and hidden passwords 1`] = `
<div>
  <main
    class="container"
  >
    <h1
      class="heading"
    >
      Set Identity Password
    </h1>
    <p
      class="subline"
    >
      You need a secure password for every Identity. A strong password looks like this: 
       
      House 12! small DOGS
    </p>
    <h2
      class="criteriaHeading"
    >
      You need to use:
    </h2>
    <ul
      class="criteria"
    >
      <li>
        uppercase and lowercase letters
      </li>
      <li>
        punctuation or special characters
      </li>
      <li>
        12 characters minimum
      </li>
      <li>
        numbers
      </li>
      <li>
        and don’t use our example
      </li>
    </ul>
    <form>
      <p
        class="inputLine"
      >
        <input
          aria-label="Please enter your password"
          autocomplete="new-password"
          class="input"
          minlength="12"
          name="password"
          placeholder="Please enter your password"
          required=""
          type="password"
        />
        <button
          aria-label="Show"
          class="show"
          title="Show"
          type="button"
        />
        <output
          class="errorTooltip"
          hidden=""
        />
      </p>
      <div
        class="complexity complexityNone"
      >
        <div
          class="lock"
        />
      </div>
      <p>
        <a
          class="cancel"
          href="/"
        >
          Cancel
        </a>
        <button
          class="submit"
          disabled=""
          type="submit"
        >
          Next Step
        </button>
      </p>
    </form>
    <button
      aria-label="Back"
      class="linkBack"
      title="Back"
      type="button"
    />
  </main>
</div>
`;

exports[`CreatePassword should allow visible and hidden passwords 2`] = `
<div>
  <main
    class="container"
  >
    <h1
      class="heading"
    >
      Set Identity Password
    </h1>
    <p
      class="subline"
    >
      You need a secure password for every Identity. A strong password looks like this: 
       
      House 12! small DOGS
    </p>
    <h2
      class="criteriaHeading"
    >
      You need to use:
    </h2>
    <ul
      class="criteria"
    >
      <li>
        uppercase and lowercase letters
      </li>
      <li>
        punctuation or special characters
      </li>
      <li>
        12 characters minimum
      </li>
      <li>
        numbers
      </li>
      <li>
        and don’t use our example
      </li>
    </ul>
    <form>
      <p
        class="inputLine"
      >
        <input
          aria-label="Please enter your password"
          autocomplete="new-password"
          class="input"
          minlength="12"
          name="password"
          placeholder="Please enter your password"
          required=""
          type="text"
        />
        <button
          aria-label="Hide"
          class="hide"
          title="Hide"
          type="button"
        />
        <output
          class="errorTooltip"
          hidden=""
        />
      </p>
      <div
        class="complexity complexityNone"
      >
        <div
          class="lock"
        />
      </div>
      <p>
        <a
          class="cancel"
          href="/"
        >
          Cancel
        </a>
        <button
          class="submit"
          disabled=""
          type="submit"
        >
          Next Step
        </button>
      </p>
    </form>
    <button
      aria-label="Back"
      class="linkBack"
      title="Back"
      type="button"
    />
  </main>
</div>
`;

exports[`CreatePassword should allow visible and hidden passwords 3`] = `
<div>
  <main
    class="container"
  >
    <h1
      class="heading"
    >
      Set Identity Password
    </h1>
    <p
      class="subline"
    >
      You need a secure password for every Identity. A strong password looks like this: 
       
      House 12! small DOGS
    </p>
    <h2
      class="criteriaHeading"
    >
      You need to use:
    </h2>
    <ul
      class="criteria"
    >
      <li>
        uppercase and lowercase letters
      </li>
      <li>
        punctuation or special characters
      </li>
      <li>
        12 characters minimum
      </li>
      <li>
        numbers
      </li>
      <li>
        and don’t use our example
      </li>
    </ul>
    <form>
      <p
        class="inputLine"
      >
        <input
          aria-label="Please enter your password"
          autocomplete="new-password"
          class="input"
          minlength="12"
          name="password"
          placeholder="Please enter your password"
          required=""
          type="password"
        />
        <button
          aria-label="Show"
          class="show"
          title="Show"
          type="button"
        />
        <output
          class="errorTooltip"
          hidden=""
        />
      </p>
      <div
        class="complexity complexityNone"
      >
        <div
          class="lock"
        />
      </div>
      <p>
        <a
          class="cancel"
          href="/"
        >
          Cancel
        </a>
        <button
          class="submit"
          disabled=""
          type="submit"
        >
          Next Step
        </button>
      </p>
    </form>
    <button
      aria-label="Back"
      class="linkBack"
      title="Back"
      type="button"
    />
  </main>
</div>
`;

exports[`CreatePassword should render 1`] = `
<div>
  <main
    class="container"
  >
    <h1
      class="heading"
    >
      Set Identity Password
    </h1>
    <p
      class="subline"
    >
      You need a secure password for every Identity. A strong password looks like this: 
       
      House 12! small DOGS
    </p>
    <h2
      class="criteriaHeading"
    >
      You need to use:
    </h2>
    <ul
      class="criteria"
    >
      <li>
        uppercase and lowercase letters
      </li>
      <li>
        punctuation or special characters
      </li>
      <li>
        12 characters minimum
      </li>
      <li>
        numbers
      </li>
      <li>
        and don’t use our example
      </li>
    </ul>
    <form>
      <p
        class="inputLine"
      >
        <input
          aria-label="Please enter your password"
          autocomplete="new-password"
          class="input"
          minlength="12"
          name="password"
          placeholder="Please enter your password"
          required=""
          type="password"
        />
        <button
          aria-label="Show"
          class="show"
          title="Show"
          type="button"
        />
        <output
          class="errorTooltip"
          hidden=""
        />
      </p>
      <div
        class="complexity complexityNone"
      >
        <div
          class="lock"
        />
      </div>
      <p>
        <a
          class="cancel"
          href="/"
        >
          Cancel
        </a>
        <button
          class="submit"
          disabled=""
          type="submit"
        >
          Next Step
        </button>
      </p>
    </form>
    <button
      aria-label="Back"
      class="linkBack"
      title="Back"
      type="button"
    />
  </main>
</div>
`;

exports[`CreatePassword should report weak password errors 1`] = `
<div>
  <main
    class="container"
  >
    <h1
      class="heading"
    >
      Set Identity Password
    </h1>
    <p
      class="subline"
    >
      You need a secure password for every Identity. A strong password looks like this: 
       
      House 12! small DOGS
    </p>
    <h2
      class="criteriaHeading"
    >
      You need to use:
    </h2>
    <ul
      class="criteria"
    >
      <li>
        uppercase and lowercase letters
      </li>
      <li>
        punctuation or special characters
      </li>
      <li>
        12 characters minimum
      </li>
      <li>
        numbers
      </li>
      <li
        class="pass"
      >
        and don’t use our example
      </li>
    </ul>
    <form>
      <p
        class="inputLine"
      >
        <input
          aria-label="Please enter your password"
          autocomplete="new-password"
          class="input"
          minlength="12"
          name="password"
          placeholder="Please enter your password"
          required=""
          type="password"
        />
        <button
          aria-label="Show"
          class="show"
          title="Show"
          type="button"
        />
        <output
          class="errorTooltip"
        >
          You need to use both uppercase and lowercase letters
        </output>
      </p>
      <div
        class="complexity complexityNone"
      >
        <div
          class="lock"
        />
      </div>
      <p>
        <a
          class="cancel"
          href="/"
        >
          Cancel
        </a>
        <button
          class="submit"
          disabled=""
          type="submit"
        >
          Next Step
        </button>
      </p>
    </form>
    <button
      aria-label="Back"
      class="linkBack"
      title="Back"
      type="button"
    />
  </main>
</div>
`;

exports[`CreatePassword should report weak password errors 2`] = `
<div>
  <main
    class="container"
  >
    <h1
      class="heading"
    >
      Set Identity Password
    </h1>
    <p
      class="subline"
    >
      You need a secure password for every Identity. A strong password looks like this: 
       
      House 12! small DOGS
    </p>
    <h2
      class="criteriaHeading"
    >
      You need to use:
    </h2>
    <ul
      class="criteria"
    >
      <li
        class="pass"
      >
        uppercase and lowercase letters
      </li>
      <li>
        punctuation or special characters
      </li>
      <li>
        12 characters minimum
      </li>
      <li>
        numbers
      </li>
      <li
        class="pass"
      >
        and don’t use our example
      </li>
    </ul>
    <form>
      <p
        class="inputLine"
      >
        <input
          aria-label="Please enter your password"
          autocomplete="new-password"
          class="input"
          minlength="12"
          name="password"
          placeholder="Please enter your password"
          required=""
          type="password"
        />
        <button
          aria-label="Show"
          class="show"
          title="Show"
          type="button"
        />
        <output
          class="errorTooltip"
        >
          You need to use at least one number
        </output>
      </p>
      <div
        class="complexity complexityPoor"
      >
        <div
          class="lock"
        />
      </div>
      <p>
        <a
          class="cancel"
          href="/"
        >
          Cancel
        </a>
        <button
          class="submit"
          disabled=""
          type="submit"
        >
          Next Step
        </button>
      </p>
    </form>
    <button
      aria-label="Back"
      class="linkBack"
      title="Back"
      type="button"
    />
  </main>
</div>
`;

exports[`CreatePassword should report weak password errors 3`] = `
<div>
  <main
    class="container"
  >
    <h1
      class="heading"
    >
      Set Identity Password
    </h1>
    <p
      class="subline"
    >
      You need a secure password for every Identity. A strong password looks like this: 
       
      House 12! small DOGS
    </p>
    <h2
      class="criteriaHeading"
    >
      You need to use:
    </h2>
    <ul
      class="criteria"
    >
      <li
        class="pass"
      >
        uppercase and lowercase letters
      </li>
      <li>
        punctuation or special characters
      </li>
      <li>
        12 characters minimum
      </li>
      <li
        class="pass"
      >
        numbers
      </li>
      <li
        class="pass"
      >
        and don’t use our example
      </li>
    </ul>
    <form>
      <p
        class="inputLine"
      >
        <input
          aria-label="Please enter your password"
          autocomplete="new-password"
          class="input"
          minlength="12"
          name="password"
          placeholder="Please enter your password"
          required=""
          type="password"
        />
        <button
          aria-label="Show"
          class="show"
          title="Show"
          type="button"
        />
        <output
          class="errorTooltip"
        >
          You need to use at least one special character
        </output>
      </p>
      <div
        class="complexity complexityMedium"
      >
        <div
          class="lock"
        />
      </div>
      <p>
        <a
          class="cancel"
          href="/"
        >
          Cancel
        </a>
        <button
          class="submit"
          disabled=""
          type="submit"
        >
          Next Step
        </button>
      </p>
    </form>
    <button
      aria-label="Back"
      class="linkBack"
      title="Back"
      type="button"
    />
  </main>
</div>
`;

exports[`CreatePassword should report weak password errors 4`] = `
<div>
  <main
    class="container"
  >
    <h1
      class="heading"
    >
      Set Identity Password
    </h1>
    <p
      class="subline"
    >
      You need a secure password for every Identity. A strong password looks like this: 
       
      House 12! small DOGS
    </p>
    <h2
      class="criteriaHeading"
    >
      You need to use:
    </h2>
    <ul
      class="criteria"
    >
      <li
        class="pass"
      >
        uppercase and lowercase letters
      </li>
      <li
        class="pass"
      >
        punctuation or special characters
      </li>
      <li>
        12 characters minimum
      </li>
      <li
        class="pass"
      >
        numbers
      </li>
      <li
        class="pass"
      >
        and don’t use our example
      </li>
    </ul>
    <form>
      <p
        class="inputLine"
      >
        <input
          aria-label="Please enter your password"
          autocomplete="new-password"
          class="input"
          minlength="12"
          name="password"
          placeholder="Please enter your password"
          required=""
          type="password"
        />
        <button
          aria-label="Show"
          class="show"
          title="Show"
          type="button"
        />
        <output
          class="errorTooltip"
        >
          You need to use at least 12 characters
        </output>
      </p>
      <div
        class="complexity complexityGood"
      >
        <div
          class="lock"
        />
      </div>
      <p>
        <a
          class="cancel"
          href="/"
        >
          Cancel
        </a>
        <button
          class="submit"
          disabled=""
          type="submit"
        >
          Next Step
        </button>
      </p>
    </form>
    <button
      aria-label="Back"
      class="linkBack"
      title="Back"
      type="button"
    />
  </main>
</div>
`;

exports[`CreatePassword should report weak password errors 5`] = `
<div>
  <main
    class="container"
  >
    <h1
      class="heading"
    >
      Set Identity Password
    </h1>
    <p
      class="subline"
    >
      You need a secure password for every Identity. A strong password looks like this: 
       
      House 12! small DOGS
    </p>
    <h2
      class="criteriaHeading"
    >
      You need to use:
    </h2>
    <ul
      class="criteria"
    >
      <li
        class="pass"
      >
        uppercase and lowercase letters
      </li>
      <li
        class="pass"
      >
        punctuation or special characters
      </li>
      <li
        class="pass"
      >
        12 characters minimum
      </li>
      <li
        class="pass"
      >
        numbers
      </li>
      <li
        class="pass"
      >
        and don’t use our example
      </li>
    </ul>
    <form>
      <p
        class="inputLine"
      >
        <input
          aria-label="Please enter your password"
          autocomplete="new-password"
          class="input"
          minlength="12"
          name="password"
          placeholder="Please enter your password"
          required=""
          type="password"
        />
        <button
          aria-label="Show"
          class="show"
          title="Show"
          type="button"
        />
        <output
          class="errorTooltip"
          hidden=""
        />
      </p>
      <div
        class="complexity complexityOk"
      >
        <div
          class="lock"
        />
      </div>
      <p>
        <a
          class="cancel"
          href="/"
        >
          Cancel
        </a>
        <button
          class="submit"
          type="submit"
        >
          Next Step
        </button>
      </p>
    </form>
    <button
      aria-label="Back"
      class="linkBack"
      title="Back"
      type="button"
    />
  </main>
</div>
`;
